include/oneapi/internal/execution_context_sycl.h

```diff
@@ -42,166 +42,175 @@ namespace interface1
-        }
+            size_t hash    = 5381;
+            const char * str = key;
+            char c;
 
-        OpenClProgramRef * getProgram()
-        {
-            return _program;
+            while (c = *str++)
+            {
+                hash = ((hash << 5) + hash) + c; /* hash * 33 + c */
+            }
+            return hash;
         }
 
-        const char* getName(services::Status * status = nullptr)
+        size_t get_id(const char* name, services::Status * status = nullptr)
         {
-            if (!_program)
+            size_t id = hash(name) % size;
+            const size_t start_id = id;
+            while (arr[id] != nullptr && strcmp(arr[id]->name.c_str(), name) != 0)
             {
-                if (status)
+                id = (id + 1) % size;
+                if (id == start_id)
                 {
-                    status->add(services::ErrorExecutionContext);
+                    setStatus(status);
+                    return 0;
                 }
-                return nullptr;
             }
-            return _program->getName();
+            return id;
         }
-    };
-
-    class KernelCacheEntry
-    {
-    private:
-        KernelPtr _kernel;
-        services::String _name;
 
     public:
-        KernelCacheEntry() : _kernel(), _name() {}
+        Map()
+        {
+            for (size_t i = 0; i < size; ++i)
+                arr[i] = nullptr;
+        }
 
-        ~KernelCacheEntry() {}
+        ~Map()
+        {
+            for (size_t i = 0; i < size; ++i)
+            {
+                delete arr[i];
+                arr[i] = nullptr;
+            }
+        }
 
-        void setKernel(KernelPtr kernel, const char *name)
+        bool contain(const char* name, services::Status * status = nullptr)
         {
-            _name = name;
-            _kernel = kernel;
+            size_t id = get_id(name, status);
+            if (status != nullptr && !status->ok())
+                return false;
+            return arr[id] != nullptr;
         }
 
-        KernelPtr getKernel()
+        void add(const char* name, services::SharedPtr<T> obj, services::Status * status = nullptr)
         {
-            return _kernel;
+            bool res = contain(name, status);
+            if (status != nullptr && !status->ok())
+                return;
+            if (!res)
+            {
+                size_t id = get_id(name, status);
+                if (status != nullptr && !status->ok())
+                    return;
+                arr[id] = new Entry();
+                arr[id]->name = name;
+                arr[id]->obj = obj;
+                return;
+            }
+            setStatus(status);
         }
 
-        const char* getName()
+        services::SharedPtr<T> get(const char *name, services::Status * status = nullptr)
         {
-            return _name.c_str();
+            size_t id = get_id(name, status);
+            if (status != nullptr && !status->ok())
```

Comment:
If the `status` is `nullptr` but `get_id` failed returning the appropriate index, this function will return arr[0]. This is unexpected. I would recommend you to create internal `Status` variable to check `get_id` results:
```cpp
Status getIdStatus;
size_t id = get_id(name, &getIdStatus);

if (status != nullptr)
    *status = getIdStatus;

if (!getIdStatus)
    return services::SharedPtr<T>();
```